src/folding/circuits/hypernova/utils.rs

```diff
@@ -0,0 +1,129 @@
+use crate::folding::circuits::utils::VecFpVar;
+use ark_ff::PrimeField;
+use ark_r1cs_std::{
+    alloc::{AllocVar, AllocationMode},
+    fields::{fp::FpVar, FieldVar},
+    ToBitsGadget,
+};
+use ark_relations::r1cs::{Namespace, SynthesisError};
+use std::{borrow::Borrow, marker::PhantomData};
+
+/// `GammaVar` is a wrapper around an `FpVar` for $\gamma$.
+#[derive(Clone)]
+pub struct GammaVar<F: PrimeField> {
+    pub value: FpVar<F>,
+}
+
+impl<F: PrimeField> AllocVar<F, F> for GammaVar<F> {
+    fn new_variable<T: Borrow<F>>(
+        cs: impl Into<Namespace<F>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        FpVar::new_variable(cs, f, mode).map(|value| Self { value })
+    }
+}
+
+impl<F: PrimeField> GammaVar<F> {
+    /// An helper to compute $\gamma^{pow}$
+    pub fn pow(&self, pow: FpVar<F>) -> Result<FpVar<F>, SynthesisError> {
+        let pow = pow.to_bits_le()?;
+        self.value.pow_le(&pow)
+    }
+}
+
+/// Gadget to compute the sum of all $\gamma^{j} \cdot eq(r_{x_j}, r_x^{\prime}) \cdot \sigma_j$.
+pub struct SumMulsGammaPowsEqSigmaGadget<F: PrimeField> {
+    _f: PhantomData<F>,
+}
+
+impl<F: PrimeField> SumMulsGammaPowsEqSigmaGadget<F> {
```

Comment:
Gadget to compute $\Sigma_{j}^{j + n} \gamma^{j} \cdot eq_eval \cdot \sigma_{j}$, where $n$ is the length of the `sigmas` vector and $j$ corresponds to `ccs.t` in `compute_c_from_sigmas_and_thetas`